### PR TITLE
feat: convert to TypeScript

### PR DESCRIPTION
**Pull request by xiaozhenliu-gg5**
Converting `redis` to TypeScript. This should be totally non-breaking.

### CODE REVIEW

**Comment by alexander-fenster** [comment]
Please do not merge until we figure out what to do with `listInstances` method which stopped being auto-paginated with the new annotations.  It's the only breaking change here.

File: package.json [hunk 1]
```diff
@@ -27,33 +27,45 @@
     "Google Cloud Memorystore for Redis API"
   ],
   "scripts": {
-    "test": "mocha",
-    "samples-test": "cd samples/ && npm link ../ && npm test && cd ../",
-    "lint": "eslint '**/*.js'",
+    "test": "c8 mocha build/test",
+    "samples-test": "cd samples/ && npm link ../ && npm install && npm test && cd ../",
+    "system-test": "mocha build/system-test",
```
Comment (bcoe) [inline package.json]: to give us a full picture of coverage, I think we should gravitate towards using `c8` for running `system-test` and `sample-tests`, codecov merges reports.

File: synth.py [hunk 2]
```diff
@@ -5,26 +5,31 @@
...
+                "package-name": f"@google-cloud/{name}"
+            },
+            proto_path=f'google/cloud/{name}/{version}',
+            extra_proto_files=["google/cloud/common_resources.proto"],
+            version=version,
+    )
     s.copy(
         library,
-        excludes=['package.json', 'README.md', 'src/index.js'])
+        excludes=['package.json', 'README.md', 'src/index.ts'])
```
Comment (bcoe) [inline synth.py]: were there manual changes in `src/index.ts`? if not we should avoid excluding it.

Comment (alexander-fenster) [reply synth.py]: Yes, we are merging two versions there. We don't do it automatically yet.

File: package.json (hunk 1 above)
Comment (alexander-fenster) [reply package.json]: OK! updated `package.json`

**Comment by alexander-fenster** [comment]
What kind of breaking change?

**Comment by bcoe** [comment]
@alexander-fenster you mentioned `listInstances` was currently a breaking change, was the goal to make it not one, or will this be a major?

**Comment by alexander-fenster** [comment]
@bcoe Ah! That one is fixed! https://github.com/googleapis/gapic-generator-typescript/pull/156 fixed that, there are no other breaking changes I'm aware of here. :)

**Comment by bcoe** [comment]
@alexander-fenster :+1: awesome, land at will; any reason to keep the blocked tag on?

**Comment by alexander-fenster** [comment]
Nope! Landing now.

**Pull request merged**